src/search/algorithms/named_vector.h

```diff
@@ -0,0 +1,103 @@
+#ifndef ALGORITHMS_NAMED_VECTOR_H
+#define ALGORITHMS_NAMED_VECTOR_H
+
+#include <string>
+#include <vector>
+
+namespace named_vector {
+/*
+  NamedVector is a vector-like collection with optional names to be associated with each element.
+  It is optimized to have minimal overhead when there are no names.
+  Any name which is not specified is assumed to be an empty string.
+ */
+template<typename T>
+class NamedVector {
+    std::vector<T> elements;
+    std::vector<std::string> names;
+public:
+    NamedVector() = default;
+    NamedVector(NamedVector<T> &&other) = default;
+    NamedVector(NamedVector<T> &other) = default;
+
+    template<typename ... _Args>
+    void emplace_back(_Args && ... __args) {
+        elements.emplace_back(std::forward<_Args>(__args) ...);
+    }
+
+    void push_back(const T &element) {
+        elements.push_back(element);
+    }
+
+    T &operator[](int index) {
```

Comment:
Please add a `const` version of this because otherwise a `const NamedVector` cannot be used for much.